Lib/test/test_os.py

```diff
@@ -227,6 +227,41 @@ def test_fdopen(self):
         self.fdopen_helper('r')
         self.fdopen_helper('r', 100)
 
+    @unittest.skipUnless(hasattr(os, 'O_DIRECT'), 'test needs os.O_DIRECT')
+    def test_odirect_readinto(self):
+        fname = support.TESTFN
+        create_file(fname, bytearray(mmap.PAGESIZE*2))
+        self.addCleanup(support.unlink, fname)
+
+        # Test one page readinto
+        fd = os.open(fname, os.O_DIRECT|os.O_RDONLY)
+        f = os.fdopen(fd, 'rb')
+        m = mmap.mmap(-1, mmap.PAGESIZE)
+        f.readinto(m)
+        f.close()
+
+        # Test two page readinto
+        fd = os.open(fname, os.O_DIRECT|os.O_RDONLY)
+        f = os.fdopen(fd, 'rb')
+        m = mmap.mmap(-1, mmap.PAGESIZE * 2)
+        f.readinto(m)
+        f.close()
```

Comment:
I would prefer to use try/finally to ensure that f is also closed in case of error. Use "with f:" (or maybe https://docs.python.org/dev/library/contextlib.html#contextlib.closing ?).